milli/src/search/new/mod.rs

```diff
@@ -0,0 +1,494 @@
+mod bucket_sort;
+mod db_cache;
+mod distinct;
+mod geo_sort;
+mod graph_based_ranking_rule;
+mod interner;
+mod limits;
+mod logger;
+pub mod matches;
+mod query_graph;
+mod query_term;
+mod ranking_rule_graph;
+mod ranking_rules;
+mod resolve_query_graph;
+mod small_bitmap;
+
+mod exact_attribute;
+// TODO: documentation + comments
+// implementation is currently an adaptation of the previous implementation to fit with the new model
+mod sort;
+// TODO: documentation + comments
+mod words;
+
+#[cfg(test)]
+mod tests;
+
+use std::collections::HashSet;
+
+use bucket_sort::{bucket_sort, BucketSortOutput};
+use charabia::TokenizerBuilder;
+use db_cache::DatabaseCache;
+use exact_attribute::ExactAttribute;
+use graph_based_ranking_rule::{Exactness, Fid, Position, Proximity, Typo};
+use heed::RoTxn;
+use interner::{DedupInterner, Interner};
+pub use logger::visual::VisualSearchLogger;
+pub use logger::{DefaultSearchLogger, SearchLogger};
+use query_graph::{QueryGraph, QueryNode};
+use query_term::{located_query_terms_from_string, LocatedQueryTerm, Phrase, QueryTerm};
+use ranking_rules::{
+    BoxRankingRule, PlaceholderQuery, RankingRule, RankingRuleOutput, RankingRuleQueryTrait,
+};
+use resolve_query_graph::{compute_query_graph_docids, PhraseDocIdsCache};
+use roaring::RoaringBitmap;
+use sort::Sort;
+use words::Words;
+
+use self::geo_sort::GeoSort;
+pub use self::geo_sort::Strategy as GeoSortStrategy;
+use self::interner::Interned;
+use crate::search::new::distinct::apply_distinct_rule;
+use crate::{AscDesc, DocumentId, Filter, Index, Member, Result, TermsMatchingStrategy, UserError};
+
+/// A structure used throughout the execution of a search query.
+pub struct SearchContext<'ctx> {
+    pub index: &'ctx Index,
+    pub txn: &'ctx RoTxn<'ctx>,
+    pub db_cache: DatabaseCache<'ctx>,
+    pub word_interner: DedupInterner<String>,
+    pub phrase_interner: DedupInterner<Phrase>,
+    pub term_interner: Interner<QueryTerm>,
+    pub phrase_docids: PhraseDocIdsCache,
+}
+
+impl<'ctx> SearchContext<'ctx> {
+    pub fn new(index: &'ctx Index, txn: &'ctx RoTxn<'ctx>) -> Self {
+        Self {
+            index,
+            txn,
+            db_cache: <_>::default(),
+            word_interner: <_>::default(),
+            phrase_interner: <_>::default(),
+            term_interner: <_>::default(),
+            phrase_docids: <_>::default(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, PartialOrd, Ord, Eq)]
+pub enum Word {
+    Original(Interned<String>),
+    Derived(Interned<String>),
+}
+
+impl Word {
+    pub fn interned(&self) -> Interned<String> {
+        match self {
+            Word::Original(word) => *word,
+            Word::Derived(word) => *word,
+        }
+    }
+}
+
+/// Apply the [`TermsMatchingStrategy`] to the query graph and resolve it.
+#[allow(clippy::too_many_arguments)]
+fn resolve_maximally_reduced_query_graph(
```

Comment:
nitpick: @dureuill attempt to document this (computes documents that meet the minimal requirements to appear inside of a search depending on the query & matching strategy & input partly computed universe)